File: livekit_flows/flow.py
from pydantic import BaseModel, Field, create_model
from typing import Optional, Type
from enum import Enum


class FieldType(str, Enum):
    STRING = "string"
    INTEGER = "integer"
    FLOAT = "float"
    BOOLEAN = "boolean"


class DataField(BaseModel):
    name: str
    type: FieldType
    description: str
    required: bool = False


class Edge(BaseModel):
    condition: str
    id: str
    target_node_id: str | None = None
    collect_data: list[DataField] = Field(default_factory=list)


class FlowNode(BaseModel):
    id: str
    name: str
    instruction: str | None = None
    static_text: str | None = None
    is_final: bool = False
    edges: list[Edge] = Field(default_factory=list)


class ConversationFlow(BaseModel):
    system_prompt: str
    initial_node: str
    nodes: list[FlowNode]

    def generate_userdata_class(self) -> Type[BaseModel]:
        field_map = {}

        for node in self.nodes:
            for edge in node.edges:
                for param in edge.collect_data:
                    field_name = param.name

                    if field_name not in field_map:
                        field_map[field_name] = param
                    else:
                        # TODO: handle conflicts
                        ...

        if not field_map:
            return create_model("FlowUserData")

        field_definitions = {}

        for param in field_map.values():
            if param.type == FieldType.STRING:
                base_type = str
            elif param.type == FieldType.INTEGER:
                base_type = int
            elif param.type == FieldType.FLOAT:
                base_type = float
            elif param.type == FieldType.BOOLEAN:
                base_type = bool
            else:
                base_type = str

            if param.required:
                field_type = base_type
                field_definition = Field(description=param.description)
            else:
                field_type = Optional[base_type]
                field_definition = Field(default=None, description=param.description)

            field_definitions[param.name] = (field_type, field_definition)

        return create_model("FlowUserData", **field_definitions)

File: editor/app/validation/flow.ts
import { object, string, array, boolean, optional, record, union, literal, pipe, minLength, number } from 'valibot'

export const FieldType = union([literal('string'), literal('integer'), literal('float'), literal('boolean')])
export const HttpMethod = union([literal('GET'), literal('POST'), literal('PUT'), literal('DELETE'), literal('PATCH')])
export const ActionTriggerType = union([literal('on_enter'), literal('on_exit')])

export const DataFieldSchema = object({
  name: string(),
  type: FieldType,
  description: string(),
  required: optional(boolean()),
})

export const ActionTriggerSchema = object({
  trigger_type: ActionTriggerType,
  action_id: string(),
})

export const CustomActionSchema = object({
  id: string(),
  name: string(),
  description: string(),
  method: HttpMethod,
  url: string(),
  headers: optional(record(string(), string())),
  body_template: optional(string()),
  timeout: optional(number()),
  store_response_as: optional(string()),
})

export const EdgeSchema = object({
  id: string(),
  condition: string(),
  target_node_id: optional(string()),
  collect_data: optional(array(DataFieldSchema)),
  actions: optional(array(ActionTriggerSchema)),
})

export const FlowNodeSchema = object({
  id: string(),
  name: string(),
  instruction: optional(string()),
  static_text: optional(string()),
  is_final: optional(boolean()),
  edges: optional(array(EdgeSchema)),
  actions: optional(array(ActionTriggerSchema)),
})

export const ConversationFlowSchema = object({
  system_prompt: pipe(string(), minLength(0)),
  initial_node: string(),
  nodes: array(FlowNodeSchema),
  actions: optional(array(CustomActionSchema)),
  environment_variables: optional(record(string(), string())),
})

export const NodePositionSchema = object({
  x: number(),
  y: number(),
})

export const FlowRecordSchema = object({
  id: string(),
  name: string(),
  system_prompt: pipe(string(), minLength(0)),
  initial_node: string(),
  nodes: array(FlowNodeSchema),
  actions: optional(array(CustomActionSchema)),
  environment_variables: optional(record(string(), string())),
})
